cmd/kube-proxy/app/server.go

```diff
@@ -1088,6 +1090,23 @@ func detectNodeIPs(ctx context.Context, rawNodeIPs []net.IP, bindAddress string)
 // required, it will wait for the node to be created.
 func getNodeIPs(ctx context.Context, client clientset.Interface, name string) []net.IP {
 	logger := klog.FromContext(ctx)
+
+	// Retrieve the Node object to avoid subsequent errors and repeats
+	opts := metav1.GetOptions{}
+	kubeletutil.FromApiserverCache(&opts)
+	_ = wait.PollUntilContextCancel(ctx, 30*time.Second, true, func(ctx context.Context) (bool, error) {
+		// Get the Node object from API server to make sure:
+		//   1) kube-proxy can reach API server
+		//   2) the Node object has been created
+		_, err := client.CoreV1().Nodes().Get(ctx, name, opts)
```

Comment:
Even coming from the other PR, I'm not really sure exactly what this PR is supposed to be doing, other than introducing a 30 second startup delay in the case where kube-proxy is started before kubelet has created the corresponding node (which we don't want). We also don't want to wait forever, because at some point (which is probably even less than 30 seconds) it is more likely that kube-proxy is misconfigured than it is that the node doesn't exist yet.

If we want to "wait for node creation", we should change the code to use a watch, no?

With some restructuring, we should be able to create an informer that watches only one node, and then use that here and in `getLocalDetectors` and with the `nodeConfig` in `ProxyServer.Run()`.

Reply:
Most important goal in this PR is to fix the bug that the kube-proxy would not get the node IP after 6 attempts and proceed with loopback address, which I am not entirely sure if it's WAI. A secondary improvement is to not request etcd for the node prematurely to lower the performance impact, which is very common if kube-proxy is running standalone (i.e. kube-proxy requesting kube-apiserver with RV being "" many times until failures).

> If we want to "wait for node creation", we should change the code to use a watch, no?

Yes, probably watch works better here. I will change to watch if you agree this is a right work.

> With some restructuring, we should be able to create an informer that watches only one node, and then use that here and in getLocalDetectors and with the nodeConfig in ProxyServer.Run().

Agreed, but I am not sure how much effort needed here for restructuring as from what I can, much initialization code rely on node info before `ProxyServer.Run()`, but I am definitely less knowledgeable on proxy code.

Reply:
Oh, no, it should not proceed with the loopback address. It should probably exit with an error.